overlord/ifacestate/handlers.go

```diff
@@ -87,11 +88,16 @@ func buildConfinementOptions(st *state.State, snapInfo *snap.Info, flags snapsta
 		return interfaces.ConfinementOptions{}, fmt.Errorf("cannot get extra mount layouts of snap %q: %s", snapInfo.InstanceName(), err)
 	}
 
+	// TODO: Look up a previous value, rather than re-checking.
+	// XXX: Probably check system key or whether the listener is running.
```

Comment:
wouldn't it make sense in this world to address this to make buildConfinementOptions a method of InterfaceManager?

Reply:
I think this is a good idea.

The problem remains, though, how to check whether prompting is enabled in `generateSystemKey()`. We can pass in a `withAppArmorPrompting bool` as a parameter to `WriteSystemKey()` from `regenerateAllSecurityProfiles()`, which in turn passes it on to `generateSystemKey()`.

The problem is that `SystemKeyMismatch()` also needs to call `generateSystemKey()`, and the former is used by `snap run`, which has no access to `InterfaceManager` so can't check the value there (assuming `features.AppArmorPrompting` is unexported, as we discussed). There might be a way to have prompting value changes register in the system key by using `AppArmorParserMtime` or something of the sort, but this feels like a hack to me, no better than leaving `features.AppArmorPrompting` exported.

Reply:
Even if we do leave `features.AppArmorPrompting` exported, we don't want to check the `ParserFeatures()` unless absolutely necessary, since it has to run the parser executable. `SystemKeyMismatch()` purges the `ParserFeatures` from the system key and uses `ParserMtime` instead, since it's modified whenever the parser features change. For `AppArmorPrompting`, we can't rely on this alone, since we need to confirm whether particular features relevant to prompting have changed. But, if we store two values in the system key (prompting supported, prompting supported&&enabled), we don't have to re-check whether prompting is supported unless either `KernelFeatures` or `ParserMtime` changes. When these do change, we will be waiting for profile regeneration anyway, so we can early return a mismatch without checking the parser features. And when writing the system key, we are already probing `ParserFeatures`, so checking `AppArmorPrompting.IsSupported()` is no extra burden.